### ISSUE TITLE
[Cache] WARNING After cache:warmup

### ISSUE THREAD

**Issue by yceruto** [symfony/symfony#23344]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4.x-dev

After:
```bash
$ git clone https://github.com/yceruto/symfony-standard -b cache_warmup_warning
$ cd symfony-standard
$ composer install --prefer-dist
$ bin/console c:c --no-warmup
$ bin/console c:w
```
Stacktrace:
```bash
...
01:22:05 WARNING   [cache] Failed to save key "%5BSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D" (array)
[
  "key" => "%5BSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D",
  "type" => "array",
  "exception" => Symfony\Component\Debug\Exception\ContextErrorException {
    -context: [
      "values" => [
        "tN2hJq4bT3:%5BSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D" => []
      ],
      "lifetime" => 0
    ],
    #message: "Warning: array_keys() expects parameter 1 to be array, boolean given",
    #code: 0,
    #file: "/home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Traits/ApcuTrait.php",
    #line: 97,
    #severity: E_WARNING,
    trace: {
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Traits/ApcuTrait.php:97: {
        : try {,
        :     return array_keys(apcu_store($values, null, $lifetime));,
        : } catch (\Error $e) {
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:216: {
        : try {,
        :     $e = $this->doSave($values, $lifetime);,
        : } catch (\Exception $e) {,
        arguments: {
          $values: [ …1],
          $lifetime: 0
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:185: {
        : ,
        :     return $this->commit();,
        : }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/ChainAdapter.php:64: {
        : ,
        : $adapter->save($item);,
        : $item->defaultLifetime = $origDefaultLifetime;,
        arguments: {
          $item: Symfony\Component\Cache\CacheItem {#1 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/ChainAdapter.php:84: {
        : while (0 <= --$i) {,
        :     $saveUp($this->adapters[$i], $item);,
        : },
        arguments: {
          Symfony\Component\Cache\Adapter\ApcuAdapter { …},
          Symfony\Component\Cache\CacheItem {#1 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/TraceableAdapter.php:40: {
        : try {,
        :     $item = $this->pool->getItem($key);,
        : } finally {,
        arguments: {
          $key: "%5BSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:91: {
        : if (!isset($this->values[$key])) {,
        :     return $this->fallbackPool->getItem($key);,
        : },
        arguments: {
          $key: "%5BSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/DoctrineProvider.php:34: {
        : {,
        :     $item = $this->pool->getItem(rawurlencode($id));,
        : ,
        arguments: {
          $key: "%5BSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
        }
      },
      /home/yceruto/bug/vendor/doctrine/cache/lib/Doctrine/Common/Cache/CacheProvider.php:78: {
        : {,
        :     return $this->doFetch($this->getNamespacedId($id));,
        : },
        arguments: {
          $id: "[Symfony\Component\Form\Form@[Annot]][1]"
        }
      },
      /home/yceruto/bug/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/CachedReader.php:193: {
        : $cacheKey = $rawCacheKey . self::$CACHE_SALT;,
        : if (($data = $this->cache->fetch($cacheKey)) !== false) {,
        :     if (!$this->debug || $this->isCacheFresh($cacheKey, $class)) {,
        arguments: {
          $id: "Symfony\Component\Form\Form@[Annot]"
        }
      },
      /home/yceruto/bug/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/CachedReader.php:82: {
        : ,
        : if (false === ($annots = $this->fetchFromCache($cacheKey, $class))) {,
        :     $annots = $this->delegate->getClassAnnotations($class);,
        arguments: {
          $rawCacheKey: "Symfony\Component\Form\Form",
          $class: ReflectionClass {#2 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Validator/Mapping/Loader/AnnotationLoader.php:48: {
        : ,
        : foreach ($this->reader->getClassAnnotations($reflClass) as $constraint) {,
        :     if ($constraint instanceof GroupSequence) {,
        arguments: {
          $class: ReflectionClass {#2 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Validator/Mapping/Loader/LoaderChain.php:57: {
        : foreach ($this->loaders as $loader) {,
        :     $success = $loader->loadClassMetadata($metadata) || $success;,
        : },
        arguments: {
          $metadata: Symfony\Component\Validator\Mapping\ClassMetadata {#3 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php:116: {
        : if (null !== $this->loader) {,
        :     $this->loader->loadClassMetadata($metadata);,
        : },
        arguments: {
          $metadata: Symfony\Component\Validator\Mapping\ClassMetadata {#3 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ValidatorCacheWarmer.php:76: {
        : if ($metadataFactory->hasMetadataFor($mappedClass)) {,
        :     $metadataFactory->getMetadataFor($mappedClass);,
        : },
        arguments: {
          $value: "Symfony\Component\Form\Form"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php:48: {
        : ,
        :     $warmer->warmUp($cacheDir);,
        : },
        arguments: {
          $cacheDir: "/home/yceruto/bug/var/cache/dev"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/CacheWarmupCommand.php:68: {
        : ,
        : $warmer->warmUp($this->getContainer()->getParameter('kernel.cache_dir'));,
        : ,
        arguments: {
          $cacheDir: "/home/yceruto/bug/var/cache/dev"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:264: {
        : } else {,
        :     $statusCode = $this->execute($input, $output);,
        : },
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#4 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#5 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:887: {
        : if ($event->commandShouldRun()) {,
        :     $exitCode = $command->run($input, $output);,
        : } else {,
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#4 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#5 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:223: {
        : $this->runningCommand = $command;,
        : $exitCode = $this->doRunCommand($command, $input, $output);,
        : $this->runningCommand = null;,
        arguments: {
          $command: Symfony\Bundle\FrameworkBundle\Command\CacheWarmupCommand { …},
          $input: Symfony\Component\Console\Input\ArgvInput {#4 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#5 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:81: {
        : ,
        :     return parent::doRun($input, $output);,
        : },
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#4 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#5 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:130: {
        :     $e = null;,
        :     $exitCode = $this->doRun($input, $output);,
        : } catch (\Exception $x) {,
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#4 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#5 …}
        }
      },
      /home/yceruto/bug/bin/console:27: {
        : $application = new Application($kernel);,
        : $application->run($input);,
        : ,
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#4 …}
        }
      }
    }
  }
]
[]
01:22:05 WARNING   [cache] Failed to fetch key "%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
[
  "key" => "%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D",
  "exception" => Symfony\Component\Debug\Exception\ContextErrorException {
    -context: [
      "key" => "%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D",
      "id" => "tN2hJq4bT3:%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D",
      "f" => Closure {
        class: "Symfony\Component\Cache\CacheItem",
        parameters: {
          $key: {},
          $value: {},
          $isHit: {}
        },
        use: {
          $defaultLifetime: 0
        },
        file: "/home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/AbstractAdapter.php",
        line: "41 to 49"
      },
      "isHit" => false,
      "value" => null
    ],
    #message: "Warning: Invalid argument supplied for foreach()",
    #code: 0,
    #file: "/home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/AbstractAdapter.php",
    #line: 141,
    #severity: E_WARNING,
    trace: {
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:141: {
        : try {,
        :     foreach ($this->doFetch(array($id)) as $value) {,
        :         $isHit = true;
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/ChainAdapter.php:80: {
        : foreach ($this->adapters as $i => $adapter) {,
        :     $item = $adapter->getItem($key);,
        : ,
        arguments: {
          $key: "%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/TraceableAdapter.php:40: {
        : try {,
        :     $item = $this->pool->getItem($key);,
        : } finally {,
        arguments: {
          $key: "%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php:91: {
        : if (!isset($this->values[$key])) {,
        :     return $this->fallbackPool->getItem($key);,
        : },
        arguments: {
          $key: "%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/DoctrineProvider.php:34: {
        : {,
        :     $item = $this->pool->getItem(rawurlencode($id));,
        : ,
        arguments: {
          $key: "%5B%5BC%5DSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D"
        }
      },
      /home/yceruto/bug/vendor/doctrine/cache/lib/Doctrine/Common/Cache/CacheProvider.php:78: {
        : {,
        :     return $this->doFetch($this->getNamespacedId($id));,
        : },
        arguments: {
          $id: "[[C]Symfony\Component\Form\Form@[Annot]][1]"
        }
      },
      /home/yceruto/bug/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/CachedReader.php:233: {
        : ,
        :     return $this->cache->fetch('[C]'.$cacheKey) >= filemtime($filename);,
        : },
        arguments: {
          $id: "[C]Symfony\Component\Form\Form@[Annot]"
        }
      },
      /home/yceruto/bug/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/CachedReader.php:194: {
        : if (($data = $this->cache->fetch($cacheKey)) !== false) {,
        :     if (!$this->debug || $this->isCacheFresh($cacheKey, $class)) {,
        :         return $data;,
        arguments: {
          $cacheKey: "Symfony\Component\Form\Form@[Annot]",
          $class: ReflectionClass {#1 …}
        }
      },
      /home/yceruto/bug/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/CachedReader.php:82: {
        : ,
        : if (false === ($annots = $this->fetchFromCache($cacheKey, $class))) {,
        :     $annots = $this->delegate->getClassAnnotations($class);,
        arguments: {
          $rawCacheKey: "Symfony\Component\Form\Form",
          $class: ReflectionClass {#1 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Validator/Mapping/Loader/AnnotationLoader.php:48: {
        : ,
        : foreach ($this->reader->getClassAnnotations($reflClass) as $constraint) {,
        :     if ($constraint instanceof GroupSequence) {,
        arguments: {
          $class: ReflectionClass {#1 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Validator/Mapping/Loader/LoaderChain.php:57: {
        : foreach ($this->loaders as $loader) {,
        :     $success = $loader->loadClassMetadata($metadata) || $success;,
        : },
        arguments: {
          $metadata: Symfony\Component\Validator\Mapping\ClassMetadata {#2 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Validator/Mapping/Factory/LazyLoadingMetadataFactory.php:116: {
        : if (null !== $this->loader) {,
        :     $this->loader->loadClassMetadata($metadata);,
        : },
        arguments: {
          $metadata: Symfony\Component\Validator\Mapping\ClassMetadata {#2 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ValidatorCacheWarmer.php:76: {
        : if ($metadataFactory->hasMetadataFor($mappedClass)) {,
        :     $metadataFactory->getMetadataFor($mappedClass);,
        : },
        arguments: {
          $value: "Symfony\Component\Form\Form"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/HttpKernel/CacheWarmer/CacheWarmerAggregate.php:48: {
        : ,
        :     $warmer->warmUp($cacheDir);,
        : },
        arguments: {
          $cacheDir: "/home/yceruto/bug/var/cache/dev"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Command/CacheWarmupCommand.php:68: {
        : ,
        : $warmer->warmUp($this->getContainer()->getParameter('kernel.cache_dir'));,
        : ,
        arguments: {
          $cacheDir: "/home/yceruto/bug/var/cache/dev"
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:264: {
        : } else {,
        :     $statusCode = $this->execute($input, $output);,
        : },
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#3 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#4 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:887: {
        : if ($event->commandShouldRun()) {,
        :     $exitCode = $command->run($input, $output);,
        : } else {,
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#3 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#4 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:223: {
        : $this->runningCommand = $command;,
        : $exitCode = $this->doRunCommand($command, $input, $output);,
        : $this->runningCommand = null;,
        arguments: {
          $command: Symfony\Bundle\FrameworkBundle\Command\CacheWarmupCommand { …},
          $input: Symfony\Component\Console\Input\ArgvInput {#3 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#4 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:81: {
        : ,
        :     return parent::doRun($input, $output);,
        : },
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#3 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#4 …}
        }
      },
      /home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:130: {
        :     $e = null;,
        :     $exitCode = $this->doRun($input, $output);,
        : } catch (\Exception $x) {,
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#3 …},
          $output: Symfony\Component\Console\Output\ConsoleOutput {#4 …}
        }
      },
      /home/yceruto/bug/bin/console:27: {
        : $application = new Application($kernel);,
        : $application->run($input);,
        : ,
        arguments: {
          $input: Symfony\Component\Console\Input\ArgvInput {#3 …}
        }
      }
    }
  }
]
[]
...
```

**Comment by chalasr** [comment]
Confirmed

**Comment by xabbuh** [comment]
What is the contents of the `$values` var inside the `ApcuTrait` which causes `apcu_store()` to return `false` instead of an array?

**Comment by yceruto** [comment]
@xabbuh According to the trace this is one of them:
```bash
"values" => [
   "tN2hJq4bT3:%5BSymfony%5CComponent%5CForm%5CForm%40%5BAnnot%5D%5D%5B1%5D" => []
],
```

Note that there is another warning too:
```bash
#message: "Warning: Invalid argument supplied for foreach()",
#code: 0,
#file: "/home/yceruto/bug/vendor/symfony/symfony/src/Symfony/Component/Cache/Adapter/AbstractAdapter.php",
#line: 141,
#severity: E_WARNING,
```

**Comment by xabbuh** [comment]
Maybe related to symfony/polyfill#94 (fixed by symfony/polyfill#95)?

**Comment by jvbussel** [comment]
I got the "Failed to save key" warning when upgrading Symfony from 3.2.9 to 3.2.10

Temporarily fix the version in composer.json to 3.2.9

**Comment by iltar** [comment]
I'm not using the polyfill, but this PR broke it for me #23091

I've done some digging and it seems to work fine using the browser, it's just the CLI that breaks. The apcu calls return false instead of an array. `apcu_store` returns false despite of `null` being passed as second argument. However, even if this returns an array, `apcu_fetch` triggers this as well.

**Comment by luispabon** [comment]
Confirming this is the case on Symfony 3.3.3.

**Comment by fabpot** [comment]
ping @nicolas-grekas

**Comment by luispabon** [comment]
Fixed in 3.3.4, thank you :+1:
